sourcecode/src-0/sourcecode/Macros.scala

```diff
@@ -0,0 +1,260 @@
+package sourcecode
+
+import scala.language.implicitConversions
+import scala.quoted._
+import scala.tasty.Reflection
+
+trait NameMacros {
+  inline implicit def generate: Name =
+    ${ Macros.nameImpl }
+}
+
+trait NameMachineMacros {
+  inline implicit def generate: Name.Machine =
+    ${ Macros.nameMachineImpl }
+}
+
+trait FullNameMacros {
+  inline implicit def generate: FullName =
+    ${ Macros.fullNameImpl }
+}
+
+trait FullNameMachineMacros {
+  inline implicit def generate: FullName.Machine =
+    ${ Macros.fullNameMachineImpl }
+}
+
+trait FileMacros {
+  inline implicit def generate: sourcecode.File =
+    ${ Macros.fileImpl }
+}
+
+trait FileNameMacros {
+  inline implicit def generate: sourcecode.FileName =
+    ${ Macros.fileNameImpl }
+}
+
+trait LineMacros {
+  inline implicit def generate: sourcecode.Line =
+    ${ Macros.lineImpl }
+}
+
+trait EnclosingMacros {
+  inline implicit def generate: Enclosing =
+    ${ Macros.enclosingImpl }
+}
+
+trait EnclosingMachineMacros {
+  inline implicit def generate: Enclosing.Machine =
+    ${ Macros.enclosingMachineImpl }
+}
+
+trait PkgMacros {
+  inline implicit def generate: Pkg =
+    ${ Macros.pkgImpl }
+}
+
+trait TextMacros {
+  inline implicit def generate[T](v: => T): Text[T] = ${ Macros.text('v) }
+  inline def apply[T](v: => T): Text[T] = ${ Macros.text('v) }
+}
+
+trait ArgsMacros {
+  inline implicit def generate: Args =
+    ${ Macros.argsImpl }
+}
+
+object Util{
+  def isSynthetic(c: Reflection)(s: c.Symbol) = isSyntheticName(getName(c)(s))
+  def isSyntheticName(name: String) = {
+    name == "<init>" || (name.startsWith("<local ") && name.endsWith(">"))
+  }
+  def getName(c: Reflection)(s: c.Symbol) = {
+    import c.given
+    s.name.trim
+      .stripSuffix("$") // meh
+  }
+}
+
+object Macros {
+
+  def actualOwner(c: Reflection)(owner: c.Symbol): c.Symbol = {
+    import c.given
+    var owner0 = owner
+    // second condition is meh
+    while(Util.isSynthetic(c)(owner0) || Util.getName(c)(owner0) == "ev") {
+      owner0 = owner0.owner
+    }
+    owner0
+  }
+
+  def nameImpl(given ctx: QuoteContext): Expr[Name] = {
+    import ctx.tasty.given
+    val owner = actualOwner(ctx.tasty)(ctx.tasty.rootContext.owner)
+    val simpleName = Util.getName(ctx.tasty)(owner)
+    '{Name(${simpleName.toExpr})}
```

Comment:
Is the reference to free terms like `Name` in these splices hygienic? We use the `${c.prefix}` incantation in the Scala 2.x implementation to ensure that, not sure if somethign similar is required in Dotty

Reply:
As far as my understanding of hygiene in Scala 2 macros goes, in Scala 3 it is fine. Scala 3 quotes are compiled on definition site wheres Scala 2 macros are compiled on expansion site. So in Scala 2, if e.g. you use `Name` instead of `${c.prefix}` and comment out `package sourcecode` in `Apply.scala` tests, you'll get a bunch of errors saying that `Name` is not found. Whereas it works with the `package sourcecode` uncommented. In Scala 3, the context of expansion is irrelevant since the quote is already compiled. @nicolasstucki maybe you can provide more precise info on this.